rpc/serve.go

```diff
@@ -0,0 +1,72 @@
+package rpc
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	"capnproto.org/go/capnp/v3"
+)
+
+// Serve serves a Cap'n Proto RPC to incoming connections.
+//
+// Serve will take ownership of bootstrapClient and release it after the listener closes.
+//
+// Serve exits with the listener error if the listener is closed by the owner.
+func Serve(lis net.Listener, bootstrapClient capnp.Client) error {
+	if !bootstrapClient.IsValid() {
+		err := errors.New("BootstrapClient is not valid")
+		return err
+	}
+	// Accept incoming connections
+	for {
+		rwc, err := lis.Accept()
+		if err != nil {
+			// Since we took ownership of the bootstrap client, release it after we're done.
+			if !bootstrapClient.IsValid() {
+				err = errors.New("the bootstrap client was already released")
+			}
+			bootstrapClient.Release()
+			return err
+		}
+
+		// For each new incoming connection, create a new RPC transport connection that will serve incoming RPC requests
+		// rpc.Options will contain the bootstrap capability
+		go func() {
+			// skip if the bootstrap client has closed since receiving the connection
+			// this can happen if the server exits while incoming connections are made
+			if bootstrapClient.IsValid() {
```

Comment:
You don't need to check this repeatedly; once at the top is plenty.

Reply:
The reason I added this was to prevent a panic due to a race condition.
When the server connection is closed and bootstrapClient is released,  the go-routine that handles the last incoming connection can still be running and invoke bootstrapClient.AddRef(). AddRef panics when this happens.
It is a bit of an edge case but might avoid a proper shutdown.